### PR TITLE
Enable scheduler to update its jobs list

### PR DESCRIPTION
**Pull request by amytai**
Previously, the scheduler only populated its jobs list once at start time. This commit enables the scheduler to receive notifications for updates (insert, update, delete) to the bgw_job table. Notifications are sent via the cache invalidation framework. Whenever the scheduler receives a notification, it re-reads the bgw_job table. For each job currently in the bgw_job table, it either instantiates new scheduler state for the job or copies over any existing scheduler state, for persisting jobs. For jobs that have disappeared from the bgw_job table, the scheduler deletes any local state it has.

Note that any updates to the bgw_job table must now go through the C, so that the cache invalidation framework in catalog.c can run. In particular, this commit includes a rudimentary API for interacting with the bgw_job table, for testing purposes. This API will be rewritten in the future.

### CODE REVIEW

**Comment by amytai** [comment]
Note: I considered defining the API for insert/delete jobs from `bgw_job` in test SQL files, since I don't think we want to pollute our public-facing API (even briefly) in order to test the scheduler functionality.

At the same time, I refrained from defining the API in test files, because I wanted the tests to be able to catch, in the future, if the modified API was incorrectly updating the `bgw_job` table (in particular, not triggering the cache invalidation callback). If the future API doesn't manage to trigger cache-invalidation, the test would still silently pass because it only calls the test API.

File: src/bgw/scheduler.c [hunk 1]
```diff
@@ -325,19 +330,124 @@ scheduled_bgw_job_start(ScheduledBgwJob *sjob, register_background_worker_callba
...
+		new_sjob = lfirst(new_ptr);
+		cur_sjob = lfirst(cur_ptr);
+
+		if (cur_sjob->job.fd.id < new_sjob->job.fd.id)
+		{
+			/*
+			 * We don't need cur_sjob anymore. Make sure to clean up the job
+			 * state. Then keep advancing cur pointer until we catch up.
+			 */
+			while (cur_sjob->job.fd.id < new_sjob->job.fd.id)
```
Comment (cevian) [inline src/bgw/scheduler.c]: why do you need this inner while. Can't you just `terminate_and_cleanup_job(cur_sjob); cur_ptr = lnext(cur_ptr); continue;`

File: src/bgw/scheduler.c [hunk 2]
```diff
@@ -325,19 +330,124 @@ scheduled_bgw_job_start(ScheduledBgwJob *sjob, register_background_worker_callba
...
+			 */
+			cur_sjob->job = new_sjob->job;
+			*new_sjob = *cur_sjob;
+
+			cur_ptr = lnext(cur_ptr);
+			new_ptr = lnext(new_ptr);
+		}
+		else if (cur_sjob->job.fd.id > new_sjob->job.fd.id)
+		{
+			/* Advance the new_job list until we catch up to cur_list */
```
Comment (cevian) [inline src/bgw/scheduler.c]: don't we need to scheduled_bgw_job_transition_state_to(lfirst(new_ptr), JOB_STATE_SCHEDULED); here?

File: src/bgw/scheduler.c [hunk 3]
```diff
@@ -325,19 +330,124 @@ scheduled_bgw_job_start(ScheduledBgwJob *sjob, register_background_worker_callba
...
+	ListCell   *new_ptr = list_head(new_jobs);
+	ListCell   *cur_ptr = list_head(cur_jobs_list);
+
+	while (cur_ptr != NULL)
 	{
-		ScheduledBgwJob *sjob = lfirst(lc);
+		ScheduledBgwJob *new_sjob;
+		ScheduledBgwJob *cur_sjob;
+
+		if (new_ptr == NULL)
```
Comment (cevian) [inline src/bgw/scheduler.c]: move this up into the while conditional?

File: src/bgw/scheduler.c [hunk 4]
```diff
@@ -325,19 +330,124 @@ scheduled_bgw_job_start(ScheduledBgwJob *sjob, register_background_worker_callba
...
+			scheduled_bgw_job_transition_state_to(lfirst(ptr), JOB_STATE_SCHEDULED);
+	}
+
+	/* Free the old list */
+	list_free(cur_jobs_list);
+	return new_jobs;
+}
+
+void
+populate_scheduled_job_tuple(ListCell *sjob_lc, Datum *values)
```
Comment (cevian) [inline src/bgw/scheduler.c]: please add a comment that this is only used by test code. Also the parameter should probably be a ScheduledBgwJob *, not a ListCell *

File: src/compat.h [hunk 5]
```diff
@@ -8,6 +8,7 @@
 #define TIMESCALEDB_COMPAT_H
 
 #include <postgres.h>
+#include <pgstat.h>
```
Comment (cevian) [inline src/compat.h]: why this change?

File: test/expected/bgw_db_scheduler.out [hunk 6]
```diff
@@ -804,3 +807,346 @@ SELECT * FROM bgw_log;
...
+)
+AS :MODULE_PATHNAME LANGUAGE C VOLATILE;
+CREATE FUNCTION verify_refresh_correct() RETURNS BOOLEAN LANGUAGE PLPGSQL AS
+$BODY$
+DECLARE
+    num_jobs INTEGER;
+    num_jobs_in_list INTEGER;
+BEGIN
+    SELECT COUNT(*) from _timescaledb_config.bgw_job INTO num_jobs;
+	select COUNT(*) from ts_test_job_refresh() JOIN _timescaledb_config.bgw_job USING (id,application_name,job_type,schedule_interval,max_runtime,max_retries,retry_period) INTO num_jobs_in_list;
```
Comment (cevian) [inline test/expected/bgw_db_scheduler.out]: shouldn't we check this for correct updates as well, not just counts?

File: src/bgw/scheduler.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (amytai) [reply src/bgw/scheduler.c]: done

Comment (amytai) [reply src/bgw/scheduler.c]: done

File: src/bgw/scheduler.c (hunk 2 above)
Comment (amytai) [reply src/bgw/scheduler.c]: good point, done. The tests don't actually catch this because, in practice, because job IDs are monotonically increasing, this should never happen

File: src/bgw/scheduler.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (amytai) [reply src/bgw/scheduler.c]: done

Comment (amytai) [reply src/bgw/scheduler.c]: done

File: src/bgw/scheduler.c (hunk 4 above)
Comment (amytai) [reply src/bgw/scheduler.c]: Added comment. I didn't make the parameter `ScheduledBgwJob`, because `ScheduledBgwJob` is a static struct defined with the *.c file. I didn't want to expose the struct (hence the helper function).

File: src/compat.h (hunk 5 above)
Comment (amytai) [reply src/compat.h]: Yeah, so `WaitLatchCompat` requires `PG_WAIT_EXTENSION`, which requires the includes.
Because this is a macro, the compiler doesn't complain in `src/compat.h`, but rather in any file in which the macro gets preprocessed. I just thought it's easier to add this includes inside `compat.h`, than in all the files that might use `WaitLatchCompat`, but I'm open to arguments against.

File: test/expected/bgw_db_scheduler.out (hunk 6 above)
Comment (amytai) [reply test/expected/bgw_db_scheduler.out]: But that's what the JOIN is? It will only join if the arguments match.

File: src/compat.h (hunk 5 above)
Comment (cevian) [reply src/compat.h]: 👍

File: src/bgw/scheduler.c [hunk 7]
```diff
@@ -325,19 +330,110 @@ scheduled_bgw_job_start(ScheduledBgwJob *sjob, register_background_worker_callba
...
+static void
+terminate_and_cleanup_job(ScheduledBgwJob *sjob)
 {
-	List	   *jobs = bgw_job_get_all(sizeof(ScheduledBgwJob), mctx);
-	ListCell   *lc;
+	if (sjob->handle != NULL)
+	{
+		TerminateBackgroundWorker(sjob->handle);
+		WaitForBackgroundWorkerShutdown(sjob->handle);
+		pfree(sjob->handle);
```
Comment (cevian) [inline src/bgw/scheduler.c]: add sjob->handle = NULL for safety?

File: src/bgw/scheduler.c [hunk 8]
```diff
@@ -325,19 +330,110 @@ scheduled_bgw_job_start(ScheduledBgwJob *sjob, register_background_worker_callba
...
+		ScheduledBgwJob *cur_sjob = lfirst(cur_ptr);
+
+		if (cur_sjob->job.fd.id < new_sjob->job.fd.id)
+		{
+			/*
+			 * We don't need cur_sjob anymore. Make sure to clean up the job
+			 * state. Then keep advancing cur pointer until we catch up.
+			 */
+			terminate_and_cleanup_job(cur_sjob);
+
```
Comment (cevian) [inline src/bgw/scheduler.c]: should we pfree cur_sjob itself here. (also below where we also do cleanup)?

File: src/bgw/scheduler.c [hunk 9]
```diff
@@ -325,19 +330,110 @@ scheduled_bgw_job_start(ScheduledBgwJob *sjob, register_background_worker_callba
...
+			scheduled_bgw_job_transition_state_to(lfirst(ptr), JOB_STATE_SCHEDULED);
 	}
-	return jobs;
+
+	/* Free the old list */
+	list_free(cur_jobs_list);
+	return new_jobs;
+}
+
+/* Only used by test code */
```
Comment (cevian) [inline src/bgw/scheduler.c]: Protect with #if DEBUG?

File: src/cache_invalidate.c [hunk 10]
```diff
@@ -17,6 +17,9 @@
 #include "extension.h"
 #include "hypertable_cache.h"
 
+/* For the bgw_job_cache_invalidate_callback() function */
```
Comment (cevian) [inline src/cache_invalidate.c]: we don't use these. get rid of comment.

File: test/expected/bgw_db_scheduler.out (hunk 6 above)
Comment (cevian) [reply test/expected/bgw_db_scheduler.out]: 👍

File: src/bgw/scheduler.c (hunk 7 above)
Comment (amytai) [reply src/bgw/scheduler.c]: done, and moved that to `worker_state_cleanup`

File: src/bgw/scheduler.c (hunk 8 above)
Comment (amytai) [reply src/bgw/scheduler.c]: No, we should not pfree here or in `worker_state_cleanup`, since the list of sjobs persists for the entire duration of the interval in between when the scheduler reloads jobs. The sjob will get freed in `list_free` below, whenever we destroy the entire list of jobs.

Comment (cevian) [reply src/bgw/scheduler.c]: then I think you need to use list_free_deep?

Comment (amytai) [reply src/bgw/scheduler.c]: Yes, good point. I misread the source and thought `list_free` also deletes `lfirst`. Fixed.
